test/test_fake_quant.py

```diff
@@ -5,143 +5,103 @@
         quant_min, quant_max = 0, 255
-        X = np.random.rand(20, 20) * 125
-        X_torch = torch.from_numpy(X).float().cuda()
-        Y = _fake_quantize_per_tensor_affine_reference(
+        X = self.to_tensor(np.random.rand(20, 20) * 125, device)
+        Y = torch.dequantize(torch.quantize_linear(X, scale, zero_point, torch.qint8))
+        Y_prime = torch.fake_quantize_per_tensor_affine(
             X, scale, zero_point, quant_min, quant_max)
-        Y_prime = _intrinsic.fq_per_tensor_affine_forward(
-            X=X_torch, scale=scale, zero_point=zero_point, quant_min=quant_min,
-            quant_max=quant_max, quant_delay=0, iter=0)
-        tolerance = 1e-6
-        np.testing.assert_allclose(Y, Y_prime.cpu(), rtol=tolerance, atol=tolerance)
+        np.testing.assert_allclose(Y, Y_prime, rtol=tolerance, atol=tolerance)
 
-    """Tests the backward method. Note that this runs the reference quantization
-    and thus the errors might be originating there."""
-    @unittest.skipIf(not torch.cuda.is_available(), 'no CUDA')
-    def test_backward_cuda(self):
+    @given(device = st.sampled_from(['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']))
+    def test_fq_module(self, device):
         np.random.seed(NP_RANDOM_SEED)
 
-        scale = 3
-        zero_point = 2
         quant_min, quant_max = 0, 255
-        X = np.random.rand(20, 20) * 125
-        Y = _fake_quantize_per_tensor_affine_reference(
-            X, scale, zero_point, quant_min, quant_max)
-        dY = Y - X  # Fake gradient
-        dX = _fake_quantize_per_tensor_affine_grad_reference(
-            X, dY, scale, zero_point, quant_min=quant_min, quant_max=quant_max)
-        X_torch = torch.from_numpy(X).float().cuda()
-        dY_torch = torch.from_numpy(dY).float().cuda()
-        dX_prime = _intrinsic.fq_per_tensor_affine_backward(
-            X=X_torch, dY=dY_torch, scale=scale, zero_point=zero_point,
-            quant_min=quant_min, quant_max=quant_max, quant_delay=0, iter=0)
-        tolerance = 1e-6
-        np.testing.assert_allclose(dX, dX_prime.cpu(), rtol=tolerance, atol=tolerance)
-
-    @unittest.skipIf(not torch.cuda.is_available(), 'no CUDA')
-    def test_numerical_consistency_cuda(self):
-        '''
-        Comparing numerical consistency between CPU quantize/dequantize op and the CUDA fake quantize op
-        '''
-        np.random.seed(NP_RANDOM_SEED)
+        X = self.to_tensor(np.random.rand(20, 20) * 125, device)
+        X.requires_grad_(True)
+        fq_module = FakeQuantize(default_qconfig, quant_min, quant_max)
+        Y_prime = fq_module(X)
+        assert fq_module.scale is not None
+        assert fq_module.zero_point is not None
+        Y = _fake_quantize_per_tensor_affine_reference(X.cpu(), fq_module.scale, fq_module.zero_point, quant_min, quant_max)
+        np.testing.assert_allclose(Y.detach().numpy(), Y_prime.detach().numpy(), rtol=tolerance, atol=tolerance)
+        Y_prime.backward(X.float())
```

Comment:
Why do we need to convert to float here?

Reply:
our kernel expects float